### PR TITLE
Fix serialization/deserialization for all types in system tablet backup

### LINKED ISSUE #37871: Backup & Restore Local Database: doubles are not exactly backed up
- NaN, +Inf, -Inf are unsupported
- Precision problem, some digits may lost

### PR DESCRIPTION
**Pull request by pixcc**
### Changelog category 

* Not for changelog (changelog entry is not required)

### Description for reviewers 

Updates the system tablet backup/restore pipeline to improve round-tripping across more column types and edge-case values (notably floating-point NaN/Inf and non-UTF8 text), and adds new unit tests to validate behavior.

### CODE REVIEW

File: ydb/core/tablet_flat/flat_executor_recovery.cpp [hunk 1]
```diff
@@ -70,11 +67,14 @@ TRawTypeValue MakeTypeValueFromJson(NScheme::TTypeInfo type, const NJson::TJsonV
...
-                throw yexception() << "Invalid value " << value << " for type " << typeId;
-            }
             return TRawTypeValue(cell->Data(), cell->Size(), typeId);
         }
+        case NScheme::NTypeIds::Date:
+            return ToRawTypeValue<NScheme::TDate>(static_cast<ui16>(value.GetUIntegerSafe()), pool);
+        case NScheme::NTypeIds::Datetime:
+            return ToRawTypeValue<NScheme::TDatetime>(static_cast<ui32>(value.GetUIntegerSafe()), pool);
+        case NScheme::NTypeIds::Timestamp:
+            return ToRawTypeValue<NScheme::TTimestamp>(value.GetUIntegerSafe(), pool);
```
[the remark this answers is not in the document]
Comment (pixcc) [reply ydb/core/tablet_flat/flat_executor_recovery.cpp:82]: Бекапы не включали нигде кроме кластеров для тестов

File: ydb/core/tablet_flat/ut/ut_backup.cpp [hunk 2]
```diff
@@ -3211,6 +3480,571 @@ Y_UNIT_TEST_SUITE(Backup) {
...
+            UNIT_ASSERT(std::isinf(*negInf.DoubleVal) && *negInf.DoubleVal < 0);
+            UNIT_ASSERT(std::isinf(*negInf.FloatVal) && *negInf.FloatVal < 0);
+
+            auto nan = env.ReadAllTypesRow(NAN_ROW_KEY);
+            UNIT_ASSERT(std::isnan(*nan.DoubleVal));
+            UNIT_ASSERT(std::isnan(*nan.FloatVal));
+
+            auto negZero = env.ReadAllTypesRow(NEGATIVE_ZERO_ROW_KEY);
+            UNIT_ASSERT(*negZero.DoubleVal == -0.0);
+            UNIT_ASSERT(*negZero.FloatVal == -0.0f);
```
[the remark this answers is not in the document]
Comment (pixcc) [reply ydb/core/tablet_flat/ut/ut_backup.cpp:3812]: -0 не переживает round-trip, восстанавливается обычный 0. Но этот тест хотя бы проверяет, что ничего не падает.

File: ydb/core/tablet_flat/flat_executor_backup.cpp [hunk 3]
```diff
@@ -98,19 +100,21 @@ void WriteColumnToJson(const TString& columnName, NScheme::TTypeId columnType,
         writer.WriteKey(columnName).WriteBool(columnData.AsValue<bool>());
         break;
     case NScheme::NTypeIds::Double:
-        writer.WriteKey(columnName).WriteDouble(columnData.AsValue<double>());
+        writer.WriteKey(columnName).WriteDouble(columnData.AsValue<double>(),
```
Comment (CyberROFL) [inline ydb/core/tablet_flat/flat_executor_backup.cpp]: Давай вот это переиспользовать: https://github.com/ydb-platform/ydb/blob/d824b4cdac6b828a55d09b813d10df6437a73736/ydb/core/tx/datashard/change_record_cdc_serializer.cpp#L104-L115

File: ydb/core/tablet_flat/flat_executor_backup.cpp [hunk 4]
```diff
@@ -98,19 +100,21 @@ void WriteColumnToJson(const TString& columnName, NScheme::TTypeId columnType,
...
+            EFloatToStringMode::PREC_NDIGITS, std::numeric_limits<double>::max_digits10);
         break;
     case NScheme::NTypeIds::Float:
-        writer.WriteKey(columnName).WriteFloat(columnData.AsValue<float>());
+        writer.WriteKey(columnName).WriteFloat(columnData.AsValue<float>(),
+            EFloatToStringMode::PREC_NDIGITS, std::numeric_limits<float>::max_digits10);
         break;
     case NScheme::NTypeIds::Date:
-        writer.WriteKey(columnName).WriteString(TInstant::Days(columnData.AsValue<ui16>()).ToString());
+        writer.WriteKey(columnName).WriteULongLong(columnData.AsValue<ui16>());
```
Comment (CyberROFL) [inline ydb/core/tablet_flat/flat_executor_backup.cpp]: Почему? cell_maker ожидает тут строки. В cdc, например, мы тут тоже пишем строки.

File: ydb/core/tablet_flat/flat_executor_backup.cpp [hunk 5]
```diff
@@ -143,11 +147,6 @@ void WriteColumnToJson(const TString& columnName, NScheme::TTypeId columnType,
             .EndList();
         break;
     }
-    case NScheme::NTypeIds::ActorId: {
```
Comment (CyberROFL) [inline ydb/core/tablet_flat/flat_executor_backup.cpp:146]: Это оказалось не нужно?

Comment (pixcc) [reply ydb/core/tablet_flat/flat_executor_backup.cpp:146]: Этот метод сериализации теряет данные. Так как в `ToString` не записывается информация о `ServiceId` и `PoolIndex`.

Это минорное и может никогда не стрельнуть, но решил сделать на всякий случай, чтобы не было никакой потери данных и сериализация соотвествовала той, что используется при записи `ActorID` в локальную базу.

Comment (CyberROFL) [reply ydb/core/tablet_flat/flat_executor_backup.cpp:146]: Кажется, Hive персистит в локальную базу ActorId. Значит мы не сможем получить для него резервную копию?

File: ydb/core/tablet_flat/flat_executor_backup.cpp (hunk 3 above)
Comment (pixcc) [reply ydb/core/tablet_flat/flat_executor_backup.cpp]: Я использую ровно эти настройки. Но этот конфиг для обертки над тем классом, что я использую в этом коде, а не для моего класса, я не могу конфиг таким образом использовать. Но фактически я использую ровно эти настройки, просто передаю их в других разных местах.

Я могу взять более высокоуровневый класс и тут провести рефакторинг, но не уверен что там будет с эксейпингом строк. Вижу в коде какие-то слова UNSAFE (Turn escaping off).

File: ydb/core/tablet_flat/flat_executor_backup.cpp (hunk 4 above)
Comment (pixcc) [reply ydb/core/tablet_flat/flat_executor_backup.cpp]: В локальную базу можно записать большую дату, так как она представима числом, например `Max<ui16>`.

При сериализации в строку дата записывается в формате ISO 8061, которое вводит ограничение на минимальное/максимальное значение даты.

Тогда получается, что в локальную базу мы смогли записать большое число, а бекап/восстановление перестало работать.

В cdc писать строки нормально, так как значения, которые пользователи могут записать в свои таблицы валидируются.

File: ydb/core/tablet_flat/flat_executor_backup.cpp (hunk 3 above)
Comment (CyberROFL) [reply ydb/core/tablet_flat/flat_executor_backup.cpp]: Мне кажется, не надо копи-пастить настройки, а положить их в одно место и всегда использовать правильную (де)сериализацию.

File: ydb/core/tablet_flat/flat_executor_backup.cpp (hunk 5 above)
Comment (pixcc) [reply ydb/core/tablet_flat/flat_executor_backup.cpp:146]: Сможем. Теперь она просто по-другому сериализуется, более "безопасным" способом, так как попадает в default case в свитче.

Сырые байты ActorID записываются в base64 формате в бекап, а потом восстанавливаются.

File: ydb/core/tablet_flat/flat_executor_recovery.cpp [hunk 6]
```diff
@@ -833,7 +825,9 @@ class TTxUploadSnapshot : public TTransactionBase<TRecoveryShard> {
 
             try {
                 NJson::TJsonValue json;
-                NJson::ReadJsonTree(line, &json, true);
+                NJson::TJsonReaderConfig config;
+                config.DontValidateUtf8 = true;
```
Comment (CyberROFL) [inline ydb/core/tablet_flat/flat_executor_recovery.cpp]: Я бы это тоже унес в io_formats/json: `DefaultWriteJsonConfig` и `DefaultReadJsonConfig`.

**Review by CyberROFL — DISMISSED** [review]

File: ydb/core/tablet_flat/flat_executor_backup.cpp [hunk 7]
```diff
@@ -55,76 +56,76 @@ EScanStatus ToScanStatus(EStatus status) {
...
-        writer.WriteKey(columnName).WriteInt(columnData.AsValue<i16>());
+        writer.Write(columnName, columnData.AsValue<i16>());
         break;
     case NScheme::NTypeIds::Uint16:
-        writer.WriteKey(columnName).WriteULongLong(columnData.AsValue<ui16>());
+        writer.Write(columnName, columnData.AsValue<ui16>());
         break;
     case NScheme::NTypeIds::Bool:
-        writer.WriteKey(columnName).WriteBool(columnData.AsValue<bool>());
+        writer.Write(columnName, columnData.AsValue<bool>());
```
[the remark this answers is not in the document]
Comment (pixcc) [reply ydb/core/tablet_flat/flat_executor_backup.cpp:99]: Uint8 и Byte указывают на одно значение, в свитче они попадают в один кейс

File: ydb/core/tablet_flat/ut/ut_backup.cpp [hunk 8]
```diff
@@ -460,6 +537,180 @@ struct TTxWriteNoBackupColumn : public ITransaction {
...
+        Result.DatetimeVal = row.GetValueOrDefault<T::DatetimeCol>();
+        Result.TimestampVal = row.GetValueOrDefault<T::TimestampCol>();
+        Result.IntervalVal = row.GetValueOrDefault<T::IntervalCol>();
+        Result.Date32Val = row.GetValueOrDefault<T::Date32Col>();
+        Result.Datetime64Val = row.GetValueOrDefault<T::Datetime64Col>();
+        Result.Timestamp64Val = row.GetValueOrDefault<T::Timestamp64Col>();
+        Result.Interval64Val = row.GetValueOrDefault<T::Interval64Col>();
+        Result.PairVal = row.GetValueOrDefault<T::PairCol>();
+        Result.ActorIdVal = row.GetValueOrDefault<T::ActorIdCol>();
+        Result.VectorVal = row.GetValueOrDefault<T::VectorCol>();
```
[the remark this answers is not in the document]
Comment (pixcc) [reply ydb/core/tablet_flat/ut/ut_backup.cpp:689]: Это обычное поведение локальной базы, если нет значения – тогда возвращать default

File: ydb/core/tablet_flat/ut/ut_backup.cpp [hunk 9]
```diff
@@ -460,6 +537,180 @@ struct TTxWriteNoBackupColumn : public ITransaction {
...
+        
+        if (ready == NTable::EReady::Page) {
+            return false;
+        } else if (ready == NTable::EReady::Data) {
+            Result.JsonDocVal = !rowState.Get(0).IsNull() ? rowState.Get(0).AsBuf() : TString();
+            Result.FloatVal = !rowState.Get(1).IsNull() ? rowState.Get(1).AsValue<float>() : 0.0f;
+            Result.Int8Val = !rowState.Get(2).IsNull() ? rowState.Get(2).AsValue<i8>() : 0;
+            Result.Int16Val = !rowState.Get(3).IsNull() ? rowState.Get(3).AsValue<i16>() : 0;
+            Result.Uint16Val = !rowState.Get(4).IsNull() ? rowState.Get(4).AsValue<ui16>() : 0;
+            Result.JsonVal = !rowState.Get(5).IsNull() ? rowState.Get(5).AsBuf() : TString();
```
[the remark this answers is not in the document]
Comment (pixcc) [reply ydb/core/tablet_flat/ut/ut_backup.cpp:703]: Это обычное поведение локальной базы, если нет значения – тогда возвращать default

File: ydb/core/io_formats/cell_maker/cell_maker.cpp [hunk 10]
```diff
@@ -268,19 +271,26 @@ namespace {
...
-        jsonConfig.ValidateUtf8 = false;
-        jsonConfig.WriteNanAsString = true;
-        return jsonConfig;
-    }
-
     TString WriteJson(const NJson::TJsonValue& json) {
         TStringStream str;
-        NJson::WriteJson(&str, &json, DefaultJsonConfig());
+        NJson::WriteJson(&str, &json, DefaultJsonWriterConfig());
         return str.Str();
```
[the remark this answers is not in the document]
Comment (pixcc) [reply ydb/core/io_formats/cell_maker/cell_maker.cpp:277]: Вызывается правильная функция из анонимного неймспейса

File: ydb/core/tablet_flat/flat_executor_recovery.cpp [hunk 11]
```diff
@@ -74,11 +72,14 @@ TRawTypeValue MakeTypeValueFromJson(NScheme::TTypeInfo type, const NJson::TJsonV
...
-                throw yexception() << "Invalid value " << value << " for type " << typeId;
-            }
             return TRawTypeValue(cell->Data(), cell->Size(), typeId);
         }
+        case NScheme::NTypeIds::Date:
+            return ToRawTypeValue<NScheme::TDate>(static_cast<ui16>(value.GetUIntegerSafe()), pool);
+        case NScheme::NTypeIds::Datetime:
+            return ToRawTypeValue<NScheme::TDatetime>(static_cast<ui32>(value.GetUIntegerSafe()), pool);
+        case NScheme::NTypeIds::Timestamp:
+            return ToRawTypeValue<NScheme::TTimestamp>(value.GetUIntegerSafe(), pool);
```
[the remark this answers is not in the document]
Comment (pixcc) [reply ydb/core/tablet_flat/flat_executor_recovery.cpp:82]: Чексуммы ловят проблемы с коррапшеном

**Review by CyberROFL — APPROVED** [review]

**Review by lex007in — APPROVED** [review]
